api_core/google/api_core/datetime_helpers.py

```diff
@@ -179,3 +179,109 @@ def to_rfc3339(value, ignore_zone=True):
         value = value.replace(tzinfo=None) - value.utcoffset()
 
     return value.strftime(_RFC3339_MICROS)
+
+
+class TimestampWithNanoseconds(datetime.datetime):
+    """Track nanosecond in addition to normal datetime attrs.
+
+    nanosecond can be passed only as a keyword argument.
+    """
+    __slots__ = ('_nanosecond',)
+
+    # pylint: disable=arguments-differ
+    def __new__(cls, *args, **kw):
+        nanos = kw.pop('nanosecond', 0)
+        if nanos > 0:
+            if 'microsecond' in kw:
+                raise TypeError(
```

Comment:
This is a `ValueError`, not a `TypeError`.

Reply:
@tseaver might have meant that it is a TypeError to use this class in that case?  I'll wait for his answer.

Reply:
Grey area, actually:  the "allowed arguments" to a constructor are actually part of the objects "type," in a formal sense:  there is nothing wrong with the *value* being passed, it is just invalid to pass both `microsecond` and `nanosecond`.